### PR TITLE
[MVCC PR] Improve forward scan

### PR DESCRIPTION
**Pull request by breeswish**
## What have you changed? (mandatory)

This PR is an improvement over the proposed [Mvcc refinement PR](https://github.com/tikv/tikv/pull/3344).

It:
- Mostly removes the usage of near_seek in forward scan
- Prevents key clone in more cases
- Adds very verbose comments

## What are the type of the changes? (mandatory)

- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)

CI

## Does this PR affect documentation (docs/docs-cn) update? (mandatory)

No

## Does this PR affect tidb-ansible update? (mandatory)

No

## Benchmark result if necessary (optional)

Later.

### CODE REVIEW

**Comment by breeswish** [comment]
@MyonKeminta @zhangjinpeng1987 PTAL

**Comment by ngaut** [comment]
Please share the bench results.

File: src/storage/mvcc/reader/forward_scanner.rs [hunk 1]
```diff
@@ -167,8 +167,28 @@ impl<S: Snapshot> ForwardScanner<S> {
             self.is_started = true;
         }
 
+        // The general idea is to simultaneously step write cursor and lock cursor.
```
Comment (zhangjinpeng1987) [inline src/storage/mvcc/reader/forward_scanner.rs]: I think you can make `read_next` much more simple when you check all locks in first call of `read_next`.

Comment (breeswish) [reply src/storage/mvcc/reader/forward_scanner.rs]: how can we gracefully handle cop limit?

File: src/storage/mvcc/reader/forward_scanner.rs [hunk 2]
```diff
@@ -180,83 +200,178 @@ impl<S: Snapshot> ForwardScanner<S> {
...
+            // ensure that the scanner keeps working for future calls even when meeting key
+            // lock errors (why??), so we don't apply `?` operator to the result here thus
+            // iterators can be moved forward.
+            let result = self.get(&current_user_key, has_write, has_lock);
 
             if has_write {
-                let next_seek_key = key.clone().append_ts(0);
-                self.write_cursor
-                    .near_seek(&next_seek_key, false, &mut self.statistics.write)?;
+                // After `self.get()`, our write cursor may be pointing to current user key (if we
```
Comment (zhangjinpeng1987) [inline src/storage/mvcc/reader/forward_scanner.rs]: You can extract following code into a function like `move_iterator_to_next_row`.

File: src/storage/types.rs [hunk 3]
```diff
@@ -162,6 +162,30 @@ impl Key {
...
+    /// Whether the user key part of a ts encoded key `ts_encoded_key` equals to the encoded
+    /// user key `user_key`.
+    #[inline]
+    #[cfg_attr(feature = "cargo-clippy", allow(cast_ptr_alignment))]
+    pub fn is_user_key_eq(ts_encoded_key: &[u8], user_key: &[u8]) -> bool {
+        let user_key_len = user_key.len();
+        if ts_encoded_key.len() != user_key_len + number::U64_SIZE {
+            return false;
+        }
+        if user_key_len >= number::U64_SIZE {
```
Comment (Connor1996) [inline src/storage/types.rs]: why check user_key last 8 bytes? seems that user_key is not a encoded key with ts.

Comment (MyonKeminta) [reply src/storage/types.rs]: `user_key` is encoded but without  ts. Here it compares `user_key` with `ts_encoded_key` 's part before ts. It's correct. But there exists other problem, that is, the `unsafe` block is really unsafe..

Comment (breeswish) [reply src/storage/types.rs]: Consider our TiDB key encoding. For example, for table data, last 8 byte is row id and the previous bytes are same. So we compare last 8 bytes first.

File: src/storage/types.rs [hunk 4]
```diff
@@ -183,12 +181,12 @@ impl Key {
         }
         if user_key_len >= number::U64_SIZE {
             // We compare last 8 bytes as u64 first, then compare the rest.
-            unsafe {
-                let ks_ptr = ts_encoded_key[user_key_len - 8..].as_ptr() as *const u64;
-                let rs_ptr = user_key[user_key_len - 8..].as_ptr() as *const u64;
-                if *ks_ptr != *rs_ptr {
-                    return false;
-                }
+            // TODO: Can we just use == to check the left part and right part? memcpy might be
```
Comment (Connor1996) [inline src/storage/types.rs]: you mean `&ts_encoded_key[user_key_len - 8..ts_encoded_key.len() - 8]  == &user_key[user_key_len - 8..]` cause memcpy?

Comment (breeswish) [reply src/storage/types.rs]: Oh, sorry, it's a typo. It should be `memcmp`.

**Comment by MyonKeminta** [comment]
Will there be benchmark results for this single PR?

File: src/storage/types.rs (hunk 4 above)
Comment (MyonKeminta) [reply src/storage/types.rs]: Please fix it

**Comment by MyonKeminta** [comment]
Rest LGTM

**Comment by breeswish** [comment]
@MyonKeminta might not. There might be a benchmark for this impl and our original MvccReader.

**Pull request merged**